### DISCUSSION TITLE
How would I get access to the embeddings per outputted segment?

### DISCUSSION THREAD

**Discussion by aamirrasheed** [openai/whisper#473] [Q&A] [answered]
I'm running audio with multiple non-overlapping speakers through Whisper, and I'd like to label every outputted segment from whisper with "Person A" "Person B" etc. The number of speakers are not known ahead of time. All audio is English.

To do this, I'd like to get access to the embeddings for every outputted segment - I plan on running that through a clustering algorithm such as k-means afterwards to differentiate between speakers.

Questions:
1. How would I get access to the embeddings per outputted segment?
2. Would you suggest a different approach for identifying speakers in the audio?

Thank you.

**Comment by jongwook** [comment, marked as answer]
Since decoding happens autoregressively and the intermediate activations are not saved after decoding each token, you can either add some code to save those activations in each forward pass or just run another round of forward pass through the decoder with your final token sequence.

But the activations of Whisper models are primarily indicative of the textual content of the speech, so a simple clustering algorithm might not give an accurate enough signal for classifying speaker identities, compared to approaches using a different speaker embedding model like #104 or #264.
